pruner/checkpoint.go

```diff
@@ -27,21 +27,44 @@ var (
 // checkpoint contains information related to the state of the
 // pruner service that is periodically persisted to disk.
 type checkpoint struct {
+	PrunerKind       string              `json:"pruner_kind"`
 	LastPrunedHeight uint64              `json:"last_pruned_height"`
 	FailedHeaders    map[uint64]struct{} `json:"failed"`
 }
 
-// DetectPreviousRun checks if the pruner has run before by checking for the existence of a
-// checkpoint.
-func DetectPreviousRun(ctx context.Context, ds datastore.Datastore) error {
-	_, err := getCheckpoint(ctx, namespace.Wrap(ds, storePrefix))
-	if errors.Is(err, errCheckpointNotFound) {
-		return nil
+func newCheckpoint(prunerKind string) *checkpoint {
+	return &checkpoint{
+		PrunerKind:       prunerKind,
+		LastPrunedHeight: 1,
+		FailedHeaders:    map[uint64]struct{}{},
 	}
+}
+
+// DetectPreviousRun ensures that a node that has been run with "full" pruning
+// mode previously cannot revert back to an "archival" one. This check should
+// only be performed when a node is either a Full or Bridge node.
+func DetectPreviousRun(ctx context.Context, ds datastore.Datastore, expectedKind string) error {
```

Comment:
Technically it's not only "detect" but also "save". Don't know what will be a better name but at least this should be mentioned in a doc-comment.